### DISCUSSION TITLE
Sync past photos + Collection Metadata: Comments from iNat email

### DISCUSSION THREAD

**Discussion by jjmontem** [rcloran/lr-inaturalist-publish#4] [General]
Hello!

@rcloran reached out on iNat, and I've just discovered this discussison forum, so I've pasted my email from that converstation:

Hi Russell,

I'm loving it so far; it's something I've been hoping was an options for a few years now! There are a couple of other plug-ins around that let you match observations with photographs in Lightroom, but it's very labor intensive and mostly manual. Your plug-in seems to be aimed at linking observations and photographs in a library and updating them with tags/IDs automatically.

So, thanks, first of all.

I'm stuggling with it a bit and I've intended to reach out. (For the record, outside of R, I have absolutely no experience coding---I'm a biologist---so I really appreciate your work and hope any comments don't sound as complaints or something). Two things so far:

1. The short synchronization wiki seems to imply that lr-inaturalist will attempt to match observations with library photographs that were not publish using the plug-in (via time stamp at the minute resolution or so). I've not seen that happen.
2. Your screenshot of the plug-in meta data seems like there are IDs and other content synchronized with photographs in the "Observations" Collection---I'm not seeing any changes to the metadata other than adding an Observation UUID and URL. The common name and other taxonomic fields remain empty. Nor are any keywords updated.

Are those two goals or expectations reasonable? I have a Mac M1 and a Windows 11 device with lightroom, so I can help test some things out if you'd like.

If there is another place this converstation would be more appropriate, let me know too!

Thanks again!

-Justin

**Comment by rcloran** [comment]
> @rcloran reached out on iNat, and I've just discovered this discussison forum, so I've pasted my email from that converstation:

You weren't missing anything, I only just set up this discussion section :) This seems like a perfect use of it.

> I'm loving it so far; it's something I've been hoping was an options for a few years now! 

Awesome! I'm glad you're finding it useful so far...

> There are a couple of other plug-ins around that let you match observations with photographs in Lightroom, but it's very labor intensive and mostly manual. Your plug-in seems to be aimed at linking observations and photographs in a library and updating them with tags/IDs automatically.

Yeah, I had looked at probably the same options as you're referring to, and built this with the goal of doing as much automatically as possible.

> So, thanks, first of all.

You're welcome, I hope you continue to find it useful.

> 1. The short synchronization wiki seems to imply that lr-inaturalist will attempt to match observations with library photographs that were not publish using the plug-in (via time stamp at the minute resolution or so). I've not seen that happen.

It will try, yes. This is fairly tricky for a few reasons. For example,

 - the data in Lightroom varies -- for example, some cameras record timezones differently, and Lightroom thus creates timestamps in different ways.
 - iNaturalist sometimes does things that make it harder to match up observations to catalog photos, like truncating the seconds from the timestamp

The main thing that would help is being able to access the photo metadata that's saved in iNaturalist. I've created a feature request for API access to photo details on the iNaturalist forums, and offered to implement it if it's approved. A vote/comment on that post might help motivate allowing me to add it: https://forum.inaturalist.org/t/photo-detail-endpoint-for-api/40481

That said, in my library I've managed to match up a pretty large percentage of my observations to photos even without it, so I think something might be going wrong with yours.

First up, there was a bug in 0.2.2 and below where the initial synchronization did not actually happen. You can force it to run by right clicking the publish service heading, and choosing `Edit Settings...`, and then clicking `Full synchronization now`. If that wasn't the reason you had no/few observations match to photos, I'd be happy to look through your logs and see if I can find what the problem might be -- please turn the log level to `Trace` in the plugin preferences (in the Lightroom Plug-in Manager) before running a full synchronization, and then send me the log file in `~/Documents/LrClassicLogs/lr-inaturalist-publish.log` (macOS) or `Documents\LrClassicLogs\lr-inaturalist-publish.log` (Windows).

> 2. Your screenshot of the plug-in meta data seems like there are IDs and other content synchronized with photographs in the "Observations" Collection---I'm not seeing any changes to the metadata other than adding an Observation UUID and URL. The common name and other taxonomic fields remain empty. Nor are any keywords updated.

Hrmm... Your understanding of what's supposed to happen sounds correct. Do you mean this is happening for photos which you have published using the plugin? Once you've published, you will need to synchronize again (no need for a full synchronize, just the incremental one will do) to receive any identifications etc that were added to the observation on iNaturalist, but ... once that link is made (ie, UUIDs exist on the photo) the synchronization should be pretty reliable -- I'm not aware of any problems with it, but it's definitely possible that there are bugs I don't yet know of. If you could let me know your process step by step, and maybe send a log of that too, that would be useful.

> Are those two goals or expectations reasonable? 

Absolutely

> I have a Mac M1 and a Windows 11 device with lightroom, so I can help test some things out if you'd like.

Thanks. I've mainly been building on macOS, and testing in a Windows 11 VM, too. There are some infuriating incompatibilities in the Lightroom API between the two, so there might be bugs on only one platform and it's more likely to be Windows since I haven't done as much testing on it. So I'm definitely interested to hear if you run into problems, especially under Windows.

**Reply by jjmontem** [reply]
Hi Russel,

> First up, there was a bug in 0.2.2 and below where the initial synchronization did not actually happen. You can force it to run by right clicking the publish service heading, and choosing Edit Settings..., and then clicking Full synchronization now. If that wasn't the reason you had no/few observations match to photos, I'd be happy to look through your logs and see if I can find what the problem might be -- please turn the log level to Trace in the plugin preferences (in the Lightroom Plug-in Manager) before running a full synchronization, and then send me the log file in ~/Documents/LrClassicLogs/lr-inaturalist-publish.log (macOS) or Documents\LrClassicLogs\lr-inaturalist-publish.log (Windows).

Full synchronization throws errors on both MacOS and Windows for me; I've attached a log from a Windows run of Full Synchronization. There is a token issue indicated---if it has to do with my log-in, the log out option is grayed out... so I can't log out. [lr-inaturalist-publish.log](https://github.com/rcloran/lr-inaturalist-publish/files/11398603/lr-inaturalist-publish.log)

> Hrmm... Your understanding of what's supposed to happen sounds correct. Do you mean this is happening for photos which you have published using the plugin? Once you've published, you will need to synchronize again (no need for a full synchronize, just the incremental one will do) to receive any identifications etc that were added to the observation on iNaturalist, but ... once that link is made (ie, UUIDs exist on the photo) the synchronization should be pretty reliable -- I'm not aware of any problems with it, but it's definitely possible that there are bugs I don't yet know of. If you could let me know your process step by step, and maybe send a log of that too, that would be useful.

Yes, I have not seen synchronizatoin of metadata beyond the UUID and URL for photographs published using Lightroom. Perhaps this has to do with the synchronization issue above. I'm able to hit publish for new photographs, synchronization occurs (inconsistently) and the observations appear on iNaturalist.

**Reply by rcloran** [reply]
I see from the logs that you're running an older version than I released yesterday -- which version are you running?

In version 0.2.0 I changed how access tokens were stored but didn't add any code to do the migration because I thought nobody was using this yet; I added some code to do the migration in 0.2.1. Do you remember which version you started with?

It's odd to me that sync does not work, but publish does -- the difference between those paths is that I put the access token migration in the sync path, but I can't figure out what mistake I might've made.

I do need to add a log out anyways, so I'll get that in either way, but for now it'd be good to know about what versions you are and were running.

**Reply by rcloran** [reply]
0.3.1-dev.4 (and newer) have log out functionality, and also remove the access token migration code I mentioned, which I'm concerned is causing a problem in your case, though I can't figure out how it might be. You can get that version from the nightly release page: https://github.com/rcloran/lr-inaturalist-publish/releases/tag/nightly

**Reply by jjmontem** [reply]
> I see from the logs that you're running an older version than I released yesterday -- which version are you running?
> 
> In version 0.2.0 I changed how access tokens were stored but didn't add any code to do the migration because I thought nobody was using this yet; I added some code to do the migration in 0.2.1. Do you remember which version you started with?
> 
> It's odd to me that sync does not work, but publish does -- the difference between those paths is that I put the access token migration in the sync path, but I can't figure out what mistake I might've made.
> 
> I do need to add a log out anyways, so I'll get that in either way, but for now it'd be good to know about what versions you are and were running.

When I ran that Full Sync attempt, I had 0.30 installed. I'm not sure why the log would say something less. 

I've installed your nightly (0.3.1-dev.4) and the full sync ran, some observations were begining to be matched with photographs, then another error occured. Running the Publish initiates a full sync again, but continues and doesn't result it in an error message output, but the log does include an error. Still, no metadata is synchronized from iNaturalist on previously published photographs within the Observations collections.
[lr-inaturalist-publish.log](https://github.com/rcloran/lr-inaturalist-publish/files/11400939/lr-inaturalist-publish.log)

**Reply by jjmontem** [reply]
Oh, and I was able to log out and log in again!

**Reply by rcloran** [reply]
> When I ran that Full Sync attempt, I had 0.30 installed. I'm not sure why the log would say something less.

Hrm. The only logging I had was that there is an update available, which makes it hard to figure out much further... I've already added (in 0.3.1-dev.4) some better logging about which version is running, so hopefully if it happens again we'll be able to diagnose things better. 

> I've installed your nightly (0.3.1-dev.4) and the full sync ran, some observations were begining to be matched with photographs, then another error occured. Running the Publish initiates a full sync again, but continues and doesn't result it in an error message output, but the log does include an error.

Yay! (kinda ;) ). I intentionally chose to ignore sync errors during publish so as not to block the publish, but perhaps a message saying it failed would be better than nothing at all.

> Still, no metadata is synchronized from iNaturalist on previously published photographs within the Observations collections. [lr-inaturalist-publish.log](https://github.com/rcloran/lr-inaturalist-publish/files/11400939/lr-inaturalist-publish.log)

Yeah, the way full sync works is that all changes are done in one transaction, so it has to get to the end and succeed.

The problem observation in your sync is this one which has no observation date at all: https://www.inaturalist.org/observations/151644982

That's something I hadn't anticipated could happen. 0.3.1-dev.5 (at https://github.com/rcloran/lr-inaturalist-publish/releases/tag/nightly again) should fix that.

**Reply by rcloran** [reply]
I've just published 0.3.1, which is the same as 0.3.1-dev.5 (except, of course, for the version number), so you can use that instead of the release from nightly if you want.

**Reply by rcloran** [reply]
> Hrm. The only logging I had was that there is an update available

Ah, that's exactly it -- the message I used in the log said "Found update", but actually it just meant that it successfully found some release on GitHub. I've committed some more changes to clarify the logging for updates so that I don't make this mistake again :)

**Reply by rcloran** [reply]
Hi @jjmontem -- can you try again with 0.4.0? Sync has been significantly improved, and so has logging. I think I've fixed all the issues that you've run into, but if not please add logs again and I'll continue trying to make this work for you.

**Reply by neilhnt** [reply]
I too am looking for an iNat API endpoint to get the original filename/timestamp/keywords etc. for a similar application.
I have a Lightroom metadata loader that pulls track names (and GPS) from Strava and sets the caption of photos taken on a particular hike, and I wanted to load iNat identification data to photos uploaded to iNat - but the matching by timestamp doesn't work (especially for photos taken in an extended sequence that are all loaded against a particular observation, and so which have different timestamps.  And resolution to the minute level seems problematic - especially with birds, where there can be several shots (of different birds) within a minute!  I'll upvote your iNat post...

**Reply by rcloran** [reply]
Unfortunately I haven't had any success having features added to iNaturalist server-side, even when I've given an example implementation -- as I've done in this case, and which, I would've thought, would help demonstrate that I'm unlikely to be a large "support" burden on their team as I learn their code base.

I'm still interested in having this added, so if you do see some willingness from the iNat folks and I'm not responding (it seems I have never got any forum updates on that thread 🤷‍♂️), please do ping me!

**Comment by jjmontem** [comment]
Hi @rcloran,

Yep; synchronization seems to work since 0.3! I had removed the undated observation you mentioned in your previous response, and that seems to have done the trick—full sync completed, meta data has been added to previously uploaded photographs and a handful of photos within my libraries have been matched to observations.

And 0.4 does seem to run more quickly.

I think I’ve run into another issue: when I view or edit the settings within the plugin, it asks if I’d like to republish previous observations/photographs. When I hit republish, the photos aren’t republished and remain in the “republishing” category within the observation collection. I can attach another log if you’d like.


--

Justin Montemarano, PhD (he/him)

Department of Biological Sciences

Colorado Mesa University

Wubben Science 211F

https://www.researchgate.net/profile/Justin_Montemarano



Chat with me on ***@***.***>!
Make an appointment<https://calendly.com/jmontemarano/studentoffice>!

________________________________

**Reply by rcloran** [reply]
Great to hear that sync is working for you now. I believe that observations without any date should be handled properly now, for what it's worth.

Unfortunately I don't think I can control the prompt for whether or not to republish, or the behaviour afterwards... that's just how Lightroom is. I have the same behaviour, I think -- pictures stay in the "Modified Photos to Re-Publish" section, but when I do hit "Publish" on the top right, they are properly updated on iNaturalist.

I always hit the "Leave As-Is" on the prompt after saving settings, because I don't want to re-upload all my photos to iNat. I am curious if that's what you really want to do, and what your use case for it is? As I said, as far as I know there's nothing I can do about it, just curious to learn more about different workflows so that I can better understand how people use this :)

**Reply by jjmontem** [reply]
I also usually hit "leave As-Is" but it seems that the observations are still categorized as needing to be re-published. Unless I can't really see _any_ reason why I'd edit an image or it's metadata and need it to be re-published, especially for all image files, since that's pretty bandwidth intensive. So, yes... I don't have a use case for that.... I just thought the plugin was trying and failing, so hoped to report it.

**Reply by rcloran** [reply]
Gotcha -- thanks for letting me know, anyways. If you want to get images out of "Modified Photo to Re-Publish" section, you can do that without re-publishing them by right clicking and selecting "Mark as Up-To-Date".

![image](https://github.com/rcloran/lr-inaturalist-publish/assets/1019877/4607dde2-237c-484f-bdc5-419a95664dfa)

**Comment by jjmontem** [comment]
Oh, brilliant; thanks!


--

Justin Montemarano, PhD (he/him)

Department of Biological Sciences

Colorado Mesa University

Wubben Science 211F

https://www.researchgate.net/profile/Justin_Montemarano



Chat with me on ***@***.***>!
Make an appointment<https://calendly.com/jmontemarano/studentoffice>!

________________________________
